crypto/rsa/rsa_gen.c

```diff
@@ -85,62 +134,179 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
+
+                for (j = 0; j < i; j++) {
+                    BIGNUM *prev_prime;
+
+                    if (j == 0)
+                        prev_prime = rsa->p;
+                    else if (j == 1)
+                        prev_prime = rsa->q;
+                    else
+                        prev_prime = sk_RSA_PRIME_INFO_value(prime_infos,
+                                                             j - 2)->r;
+
+                    if (!BN_cmp(prime, prev_prime)) {
+                        goto redo;
+                    }
+                }
+            }
+            if (!BN_sub(r2, prime, BN_value_one()))
+                goto err;
+            if (!BN_gcd(r1, r2, rsa->e, ctx))
+                goto err;
+            if (BN_is_one(r1))
+                break;
+            if (!BN_GENCB_call(cb, 2, n++))
+                goto err;
+        }
+
+        bitse += bitsr[i];
+
+        /* calculate n immediately to see if it's sufficient */
+        if (i == 1) {
+            /* we get at least 2 primes */
+            if (!BN_mul(r1, rsa->p, rsa->q, ctx))
+                goto err;
+        } else if (i != 0) {
+            /* modulus n = p * q * r_3 * r_4 ... */
+            if (!BN_mul(r1, rsa->n, prime, ctx))
+                goto err;
+        } else {
+            /* i == 0, do nothing */
+            if (!BN_GENCB_call(cb, 3, i))
+                goto err;
+            continue;
+        }
+        /*
+         * if |r1|, product of factors so far, is not as long as expected
+         * (by checking the first 4 bits are less than 0x9 or greater than
+         * 0xF). If so, re-generate the last prime.
+         *
+         * NOTE: This actually can't happen in two-prime case, because of
+         * the way factors are generated.
+         *
+         * Besides, another consideration is, for multi-prime case, even the
+         * length modulus is as long as expected, the modulus could start at
+         * 0x8, which could be utilized to distinguish a multi-prime private
+         * key by using the modulus in a certificate. This is also covered
+         * by checking the length should not be less than 0x9.
+         */
+        if (!BN_rshift(r2, r1, bitse - 4))
             goto err;
-    }
-    if (!BN_GENCB_call(cb, 3, 0))
-        goto err;
-    for (;;) {
-        do {
-            if (!BN_generate_prime_ex(rsa->q, bitsq, 0, NULL, NULL, cb))
+        bitst = BN_get_word(r2);
+
+        if (bitst < 0x9 || bitst > 0xF) {
+            /*
+             * For keys with more than 4 primes, we attempt longer factor to
+             * meet length requirement.
+             *
+             * Otherwise, we just re-generate the prime with the same length.
+             *
+             * This strategy has the following goals:
+             *
+             * 1. 1024-bit factors are effcient when using 3072 and 4096-bit key
+             * 2. stay the same logic with normal 2-prime key
+             */
+            if (primes > 4) {
```

Comment:
So I have tried using you patch, to run "openssl genrsa -primes +3 1024"
BTW: was this discussed why 4-prime RSA are generated with +3 ?
I was able to observe one time when the command took very much longer than normal,
and finally produced the following RSA key:
```
prime1:
    00:ca:b0:4c:07:04:64:40:71:fe:fa:8e:6c:1d:ee:
    5c:e0:5d:82:c9:cd:99:3a:06:51:7d:b1:42:fc:c1:
    f1:53:3b
prime2:
    00:c3:a7:ca:ae:94:ab:09:8c:78:aa:1a:5e:04:fb:
    f0:e0:6d:38:13:3f:fd:85:7e:fc:b8:14:28:66:74:
    8b:f0:ff
prime3:
    00:ed:fd:fb:44:87:12:28:d4:28:78:50:bf:c0:07:
    14:12:9b:be:47:ba:9c:ff:d2:6f:48:2d:88:df:2a:
    07:55:9d
prime4:
    00:ff:fb:bd:bb:3a:19:30:8f:eb:f0:26:34:4d:85:
    76:d4:d6:85:a7:4b:8a:0a:69:57:2e:b6:20:0c:ab:
    72:ca:85
```

Reply:
```
p = 0xcab04c0704644071fefa8e6c1dee5ce05d82c9cd993a06517db142fcc1f1533b
q = 0xc3a7caae94ab098c78aa1a5e04fbf0e06d38133ffd857efcb8142866748bf0ff
r = 0xedfdfb44871228d4287850bfc00714129bbe47ba9cffd26f482d88df2a07559d
s = 0xfffbbdbb3a19308febf026344d8576d4d685a74b8a0a69572eb6200cab72ca85
print hex(p*q*r)
print hex(p*q*r*s)
```
The problem that p * q * r is 0x90038637dca...
and the final module is 0x900120e22af...
and it took many attempts to find a prime4 with so many
MSB at 1.

Would it be possible to set i=0 after maybe 4 unsuccessful tries,
in order to start a completely new search ?

Reply:
> BTW: was this discussed why 4-prime RSA are generated with +3 ?

please disregard that, it happened with -prime +4, so that's fine.

Reply:
> Would it be possible to set i=0 after maybe 4 unsuccessful tries,
in order to start a completely new search ?

Does this really help?

From my point of view, I think we could make this 'avoid-characteristic-of-multi-prime' as a option for the user to choose, for instance provide a flag and a corresponding function to set this flag. If a user wants to do so, then he/she might suffer a longer key generating process occasionally. If the flag is not set, the key generating would be faster but the modulus might be more distinguishable either.

Reply:
Well, if several times trying to generate the 4th prime number does not yield
a satisfactory result, then likelihood is that one of the previous primes was
actually to blame.  And the result will not be really nice, if you go several 100 times
in the loop, as it happened when I tried this.

Reply:
Done 'retry'